bundles/org.openhab.core.automation/src/main/java/org/openhab/core/automation/internal/RuleExecutionSimulator.java

```diff
@@ -0,0 +1,153 @@
+import java.util.Date;
+import java.util.List;
+import java.util.stream.Stream;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.openhab.core.automation.Condition;
+import org.openhab.core.automation.Rule;
+import org.openhab.core.automation.RuleExecution;
+import org.openhab.core.automation.RulePredicates;
+import org.openhab.core.automation.RuleRegistry;
+import org.openhab.core.automation.Trigger;
+import org.openhab.core.automation.handler.ConditionHandler;
+import org.openhab.core.automation.handler.TimeBasedConditionHandler;
+import org.openhab.core.automation.handler.TimeBasedTriggerHandler;
+import org.openhab.core.automation.handler.TriggerHandler;
+import org.openhab.core.scheduler.SchedulerTemporalAdjuster;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Simulates {@link Rule rules} to show up their next execution times within the schedule.
+ *
+ * @author Sönke Küper - Initial contribution
+ */
+@NonNullByDefault
+final class RuleExecutionSimulator {
+
+    /**
+     * Tag that all rules must have, do be included within the simulation.
+     */
+    private static final String TAG_SCHEDULE = "Schedule";
+
+    private final Logger logger = LoggerFactory.getLogger(RuleExecutionSimulator.class);
+
+    private final RuleRegistry ruleRegistry;
+    private final RuleEngineImpl ruleEngine;
+
+    /**
+     * Constructs an new {@link RuleExecutionSimulator}.
+     */
+    public RuleExecutionSimulator(RuleRegistry ruleRegistry, RuleEngineImpl ruleEngine) {
+        this.ruleRegistry = ruleRegistry;
+        this.ruleEngine = ruleEngine;
+    }
+
+    /**
+     * Simulates the execution of all rules with tag 'Schedule' for the given time interval.
+     * The result is sorted ascending by execution time.
+     *
+     * @param untilDate {@link Date} until the executions should be simulated.
+     * @return A {@link Stream} with all expected {@link RuleExecution}.
+     */
+    public Stream<RuleExecution> simulateRuleExecutions(ZonedDateTime fromDate, ZonedDateTime untilDate) {
+        logger.debug("Simulating rules from {} until {}.", fromDate, untilDate);
+
+        return ruleRegistry.stream() //
+                .filter(RulePredicates.hasAllTags(TAG_SCHEDULE)) //
+                .filter((Rule r) -> ruleEngine.isEnabled(r.getUID())) //
+                .map((Rule r) -> simulateExecutionsForRule(r, fromDate, untilDate)) //
+                .flatMap(List::stream).sorted();
+    }
+
+    /**
+     * Simulates the next executions for the given {@link Rule} until the given {@link Date}.
+     *
+     * @param rule {@link Rule} to be simulated.
+     * @param fromDate {@link Date} from when the executions should be simulated.
+     * @param untilDate {@link Date} until the executions should be simulated.
+     * @return List of expected {@link RuleExecution}.
+     */
+    private List<RuleExecution> simulateExecutionsForRule(Rule rule, ZonedDateTime fromDate, ZonedDateTime untilDate) {
```

Comment:
I am wondering if this method should be part of the public API too. It seems to make sense to just simulate a single rule too. We maybe can add a dedicated REST endpoint too.

Reply:
I prefer the YAGNI principle, its not needed at the moment, so if this changes it can easy be made accessible.